[CPU] Add new attribute to control peeling

banach-space opened the pull request:
Adds a new CPU lowering config attribute, `enable_loop_peeling`,
to control whether to run the loop peeling pass before vectorisation.

Peeling is highly beneficial when using "scalable" vectorisation - it
helps to get rid of masking:
  * https://discourse.llvm.org/t/psa-scalable-auto-vec-in-linalg-without-masking/

Similarly to other Ops (most notably `linalg.matmul`), peeling is
enabled through "pre-processing". It's "off" by default and can be
enabled with:
  * --iree-llvmcpu-vector-pproc-strategy=peel

banach-space committed:
Create new CPU lowering expert (conv + peeling)

Adds a new CPU lowering expert: `CPUConvTileAndDecomposeAndPeelExpert`.
This is very similar to `CPUConvTileAndDecomposeExpert`, with the
addition that peeling is run before vectorisation.

Peeling is highly beneficial when using "scalable" vectorisation - it
helps to get rid of masking:
  * https://discourse.llvm.org/t/psa-scalable-auto-vec-in-linalg-without-masking/

Similarly to other Ops (most notably `linalg.matmul`), peeling is
enabled through "pre-processing". It's "off" by default and can be
enabled with:
  * --iree-llvmcpu-vector-pproc-strategy=peel

MaheshRavishankar reviewed (changes requested):
I see why you have it this way, but it seems a bit off that we add a new pipeline just to chose a different vectorization strategy. I think it might be better to maybe update the translation info and add the vectorization stragety to that. The translation info has a filed of just dictionary attributes that could be anything. You just need to set that and look for it when you set the vector strategy. So its is a fairly localized change.

banach-space committed:
fixup! Create new CPU lowering expert (conv + peeling)

Delete `CPUConvTileAndDecomposeAndPeelExpert`, use pipeline config
attribute instead.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -2157,9 +2163,22 @@ static LogicalResult setConvRootConfig(mlir::FunctionOpInterface entryPointFn,
+    auto peelAttrVal = BoolAttr::get(ctx, true);
+    attrs.push_back({peelAttrName, peelAttrVal});
+    pipelineConfig = DictionaryAttr::get(ctx, attrs);
+  }
```
I'm confused. I thought the VectorPreProcStrategy is None for all the convolutions in `getVectorPreProcStrategy`, but we have a special logic here. Is it because we can set the strategy to peeling through CLI flags?

I can see the value about what Mahesh suggested. We decided the strategy in `KernelDispatch.cpp`, and we want to "recover" the information when adding passes to pipeline. They could conflict. It seems to be a separate issue from enabling scalable vectors for convolutions. So there are code health issue and feature support to me.

RE code health issue, I wonder if we can set the strategy on attribute in getVectorPreProcStrategy? This way we don't have floating logics everywhere to serve the need. It will need more work to make it happen cleanly, so I think it could be done in a follow-up.

RE convolution enablement: I wonder if we can retire `addConvTileAndDecomposeExpertPassPipeline`. Instead, we just use `addMultiTilingExpertPassPipeline` for convolutions as well. To achieve it, we need to add `DecomposeConvolutionToLowerDimOpsPass` to the multi-tiling pipeline, and have an additional option for `splitVectorTransfersTo`.

banach-space replied:
Thanks for taking a look - lots of great comments 🙏🏻 

>  I thought the VectorPreProcStrategy is None for all the convolutions in getVectorPreProcStrategy, but we have a special logic here. Is it because we can set the strategy to peeling through CLI flags?

Yes. 

We have no heuristics to select a pre-processing strategy for convs. With this patch, we would allow one to be set by the user. At some point, we can also turn peeling "ON" by default for SVE. 

Btw, I always use `getVectorPreProcStrategy` to retrieve the pre-processing strategy rather than reading [clPProcStrategy](https://github.com/iree-org/iree/blob/ae64f1d4b4988f2d9dcddaaf6a21e0f61d0af411/compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp#L128-L143) directly - this way we can avoid duplicating the logic to deal with `clPProcStrategy".

> We decided the strategy in KernelDispatch.cpp, and we want to "recover" the information when adding passes to pipeline.

Yes - we need to "forward" information:
*  from [setConvRootConfig](https://github.com/iree-org/iree/blob/ae64f1d4b4988f2d9dcddaaf6a21e0f61d0af411/compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp#L2109-L2112),
*  to [LLVMCPULowerExecutableTargetPass::runOnOperation()](https://github.com/iree-org/iree/blob/ae64f1d4b4988f2d9dcddaaf6a21e0f61d0af411/compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp#L88).

I see two options for this:
1. Through a dedicated `DispatchLoweringPassPipeline` that can be used to specialise in this [switch stmt](https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp#L114) - that's what I tried initially,
2. Through attributes - this is what the current implementation does.

I see pros and cons for both. You have much more experience with this, so I will happily follow your suggestion.

>  They could conflict

Can you expand? Not sure I know what you meant here.

> RE code health issue, I wonder if we can set the strategy on attribute in getVectorPreProcStrategy? 

I see what you mean. There's `CPUDoubleTilingExpert` and `CPUDoubleTilingPeelingExpert`, which are virtually identical: https://github.com/iree-org/iree/blob/ae64f1d4b4988f2d9dcddaaf6a21e0f61d0af411/compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp#L128-L139

So, if the "design" is to limit the number of lowering expert, then those two should probably be merged?

>  It will need more work to make it happen cleanly, so I think it could be done in a follow-up.

I'm happy to refactor things, but we should definitely defer that to a separate PR. I'm worried that this can lead to some surprises. I'll be under less pressure if once convs are fast for SVE :) 

>  I wonder if we can retire addConvTileAndDecomposeExpertPassPipeline. Instead, we just use addMultiTilingExpertPassPipeline for convolutions as well. 

Possibly. but looks more complex than merging `CPUDoubleTilingExpert` and `CPUDoubleTilingPeelingExpert`:
* [addMultiTilingExpertPassPipeline](https://github.com/iree-org/iree/blob/ae64f1d4b4988f2d9dcddaaf6a21e0f61d0af411/compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp#L349-L423),
* [addConvTileAndDecomposeExpertPassPipeline](https://github.com/iree-org/iree/blob/ae64f1d4b4988f2d9dcddaaf6a21e0f61d0af411/compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp#L425-L480)

These two have diverged a fair bit.

### Summary
There's quite a lot in here, so let me summarise. Please correct me if I got sth wrong or you disagree with the proposed approach.
1. Use **attributes** rather than a **new lowering expert** for this change,
2. In a follow-up PR, merge  `CPUDoubleTilingExpert` and `CPUDoubleTilingPeelingExpert` (this should "just work" ™️ )
3. In a follow-up PR, try merging `addConvTileAndDecomposeExpertPassPipeline` and `addMultiTilingExpertPassPipeline` (and remove `CPUConvTileAndDecomposeExpert`?). 

Basically, I'd like to land this one without 2. and 3. I'm less confident about 3., but am happy to work on both and see how it goes.

hanhanW replied:
Thanks for the good summary. It's aligned with what I'm thinking. Yes, we should go with (1) -> (2) and perhaps (3) later.

MaheshRavishankar reviewed (commented):
Thanks @banach-space . I think this is better, but I would wait for more feedback from Hanhan and Benoit who work on this more.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -2157,9 +2163,22 @@ static LogicalResult setConvRootConfig(mlir::FunctionOpInterface entryPointFn,
+    // This will later be extracted by LLVMCPULowerExecutableTargetPass.
+    auto ctx = convOp.getContext();
+    SmallVector<NamedAttribute> attrs;
+    auto peelAttrName = StringAttr::get(ctx, "enable_peeling");
```
Instead of having a magic string floating in different files, how about adding a `getPeelAttrName` to CPUUtils?

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -2157,9 +2163,22 @@ static LogicalResult setConvRootConfig(mlir::FunctionOpInterface entryPointFn,
+    SmallVector<NamedAttribute> attrs;
+    auto peelAttrName = StringAttr::get(ctx, "enable_peeling");
+    auto peelAttrVal = BoolAttr::get(ctx, true);
+    attrs.push_back({peelAttrName, peelAttrVal});
```
nit: emplace_back because we intend to call the constructor.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp:
```diff
@@ -140,6 +140,16 @@ void LLVMCPULowerExecutableTargetPass::runOnOperation() {
+      if (peelAttr && peelAttr->getValue() == trueAttr) {
+        pipelineOpts.enablePeeling = true;
+      }
+    }
```
Can we create a function in CPUUtils, and reuse it here? It makes codes easier to read.

banach-space committed:
fixup! fixup! Create new CPU lowering expert (conv + peeling)

Address PR comments from @hanhanW

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -211,6 +211,11 @@ getVectorPreProcStrategy(linalg::LinalgOp linalgOp) {
     return clPProcStrategy;
   }
 
+  // TODO: Implement heurystics for Convs
```
Nit: type on heuristics.

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -2157,9 +2163,22 @@ static LogicalResult setConvRootConfig(mlir::FunctionOpInterface entryPointFn,
   return setOpConfigAndEntryPointFnTranslation(
       entryPointFn, convOp, tileSizes, scalableTileFlags,
-      DispatchLoweringPassPipeline::CPUConvTileAndDecomposeExpert);
+      DispatchLoweringPassPipeline::CPUConvTileAndDecomposeExpert, {}, {},
```
Nit: please add `/*{argument_name}=*/` to the values that are set to default `{}`

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp:
```diff
@@ -63,4 +64,25 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
+  return StringAttr::get(ctx, "enable_loop_peeling");
+}
+
+bool isLoopPeelingEnabled(FunctionOpInterface *funcOp) {
```
This does not need to be a `FunctionOpInterface *`. Thats not the norm. Usually just pass these by value.

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp:
```diff
@@ -63,4 +64,25 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
 }
 
+StringAttr getPeelAttrName(MLIRContext *ctx) {
+  return StringAttr::get(ctx, "enable_loop_peeling");
```
Nit: Define a static char array with the name an use that.
```
static const char[] kLoopPeeling = "enable_loop_peeling";
```
and use `kLoopPeeling` here.

MaheshRavishankar reviewed (approved):
I am unblocking, but really would be good to wait for Hanhan's comments.

MacDue commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -2157,9 +2163,22 @@ static LogicalResult setConvRootConfig(mlir::FunctionOpInterface entryPointFn,
+    auto peelAttrName = getPeelAttrName(ctx);
+    auto peelAttrVal = BoolAttr::get(ctx, true);
+    attrs.emplace_back(peelAttrName, peelAttrVal);
+    pipelineConfig = DictionaryAttr::get(ctx, attrs);
```
nit: I think this could be clearer with some naming tweaks + a few less steps:
```c++
    auto context = convOp.getContext();
    auto enableLoopPeelingAttrName = getEnableLoopPeelingAttrName(context);
    auto trueValue = BoolAttr::get(context, true);
    pipelineConfig = DictionaryAttr::get(
        context,
        ArrayRef<NamedAttribute>({enableLoopPeelingAttrName, trueValue}));
```

banach-space replied:
Thanks, this is much nicer! I thought that I needed a `SmallVector` here, but this makes me realise that we only need storage for the actual  contents.

MacDue commented on compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp:
```diff
@@ -63,4 +64,25 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
   return rootOperation;
 }
 
+StringAttr getPeelAttrName(MLIRContext *ctx) {
```
nit: This is `getEnableLoopPeelingAttrName()`

MacDue commented on compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp:
```diff
@@ -63,4 +64,25 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
+    if (peelAttr && peelAttr->getValue() == trueAttr) {
+      return true;
+    }
+  }
```
nit: early exists (less nesting)
```c++
if (!config)
   return false;
```

MacDue replied:
(similar naming nits here this is `enableLoopPeelingAttrName`)

banach-space replied:
That's much cleaner, thanks!

MacDue reviewed (commented):
a few nits:

banach-space committed:
fixup! fixup! fixup! Create new CPU lowering expert (conv + peeling)

Address comments from Mahesh and Ben

hanhanW commented on compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp:
```diff
@@ -63,4 +66,28 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
+    return false;
+  }
+
+  auto peelAttr = translationInfo.getConfiguration().getNamed(peelAttrName);
```
nit: Use config directly, i.e., `config.getNamed(...)`.

hanhanW commented on compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp:
```diff
@@ -63,4 +66,28 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
+
+bool isLoopPeelingEnabled(FunctionOpInterface funcOp) {
+  auto peelAttrName = getEnableLoopPeelingAttrName(funcOp->getContext());
+  auto trueAttr = BoolAttr::get(funcOp->getContext(), true);
```
Do we use `UnitAttr` instead? The attribute name is `enable_loop_peeling`. It is clear that we want peeling when the attribute is present. What do you think?

MacDue replied:
+1 to a unit attribute

hanhanW commented on compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp:
```diff
@@ -63,4 +66,28 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
+
+  if (config) {
+    return false;
+  }
```
Should this be `!config`?

banach-space replied:
Yeah, realised that after pushing.

MacDue commented on compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp:
```diff
@@ -63,4 +66,28 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
+    return false;
+  }
+
+  auto peelAttr = translationInfo.getConfiguration().getNamed(peelAttrName);
```
nit: `peelAttr` -> `enableLoopPeelingAttr` 
nit:  `enableLoopPeelingAttrName`

MacDue commented on compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp:
```diff
@@ -63,4 +66,28 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
+  }
+
+  return false;
+}
```
Not relevant if using a `UnitAttr`, but `BoolAttr::getValue()` returns a normal `bool` (there's no need to construct a `trueAttr` attribute). Also, the `if` is redundant here. 
```suggestion
  return peelAttr && peelAttr->getValue();
}
```

banach-space committed:
fixup! fixup! fixup! fixup! Create new CPU lowering expert (conv + peeling)

Switch to UnitAttr, rename vars

MacDue commented on compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp:
```diff
@@ -63,4 +66,22 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
+
+  auto enableLoopPeelingAttrName = getEnableLoopPeelingAttrName(context);
+  auto enableLoopPeelingAttr = config.getNamed(enableLoopPeelingAttrName);
+  return enableLoopPeelingAttr.has_value() ? true : false;
```
nit: There's no need to check a unit attribute (simply having the attribute enables peeling). Also, you don't need to construct a `StringAttr` to check (`.contains` can take a `StringRef`).
```suggestion
  return config.contains(kLoopPeelingAttrName);
```

MacDue replied:
nit I just noticed: `kLoopPeelingAttrName` -> `kEnableLoopPeelingAttrName`

MacDue reviewed (approved):
Final nit, otherwise LGTM.

MacDue commented on compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp:
```diff
@@ -63,4 +66,22 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
+
+  auto enableLoopPeelingAttrName = getEnableLoopPeelingAttrName(context);
+  auto enableLoopPeelingAttr = config.getNamed(enableLoopPeelingAttrName);
+  return enableLoopPeelingAttr.has_value() ? true : false;
```
Or or even just:
```suggestion
  return config && config.contains(kLoopPeelingAttrName);
```

MacDue replied:
(no need for the `auto context = funcOp->getContext();` too)

banach-space committed:
fixup! fixup! fixup! fixup! fixup! Create new CPU lowering expert (conv + peeling)

Simplify

hanhanW commented on compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp:
```diff
@@ -63,4 +66,18 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
+    return false;
+  }
+
+  return config.contains(kLoopPeelingAttrName);
```
nit: I think we can collapse it to `return config && config.contains(...);`. What do you think?

banach-space replied:
Ben suggested exactly that and in my head I've already added that 🤦🏻 But that was imaginary code :) Thanks for pointing this out, sending update shortly.

hanhanW reviewed (approved):
Thanks for addressing all the comments! Just a final nit, please take a look.

banach-space committed:
fixup! fixup! fixup! fixup! fixup! fixup! Create new CPU lowering expert (conv + peeling)

simplify
